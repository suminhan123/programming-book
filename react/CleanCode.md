## Clean Code 란?

출처 : https://www.youtube.com/watch?v=edWbHp_k_9Y 를 보고 정리해서 글을 작성하였습니다.

클린 코드라 하면 '명확한 이름', '중복 줄이기'를 먼저 말한다.
실무에선 이 외에도 조금 더 섬세하게 코드를 정리하는 스킬이 필요하다.

### 1. 실무에서 클린 코드의 의의

"그 코드는 안건드리시는 게 좋을거에요. 일단 제가 만질게요.^^;;"
=> 흐름 파악이 어렵고, 도메인 맥락이 표현이 안되어 동료에게 물어봐야 할 수 있는 코드

=> 이런 코드는 개발할 때 병목되고, 유지 보수할 때 시간이 오려걸리며 심하면 기능 추가가 불가능한 상태

따라서, 클린 코드는 **유지 보수의 시간을 단축**시킬 수 있다. (코드 파악이 쉽고, 디버깅 및 리뷰하기 좋다)

실무에서 **기존 코드에 기능을 추가하는 상황**이 대다수이다! 어떻게 기존코드에 기능을 추가할 수 있을까?

### 2. 안일한 코드 추가의 함정

코드를 안일하게 추가하면 함정에 빠질 수 있다.
다음은 예제입니다.
![](https://velog.velcdn.com/images/hansoom3315/post/463fa2a4-ed84-4781-bfb8-189197db3db7/image.png)
보험에 대한 질문을 입력하는 페이지가 있는데 내 설계사가 있는 경우에 그 설계사 사진이 들어간 팝업을 먼저 띄워달라는 기능 추가이다.

- 기존 코드

```javascript
function QuestionPage() {
  async function handleQuestionSubmit() {
    // 약관 동의 필요하면 팝업 띄우기
    const 약관동의 = await 약관동의_받아오기();
    if (!약관동의) {
      await 약관동의_팝업열기();
    }
    // 질문을 전송하고 선공했다 alert 창을 띄움
    await 질문전송(questionValue);
    alert("질문이 등록되었어요.");
  }

  return (
    <form>
      <textarea placeholder="어떤내용이 궁금하신가요?" />
      //질문하기 버튼 클릭
      <Button onClick={handleQuestionSubmit} />
    </form>
  );
}
```

이 코드에 새 기능을 추가해보자

아래 코드와 같이, onClick 핸들러에 연결 중인 전문가가 있으면 팝업 띄우는 로직을 추가하고 해당 팝업 컴포넌트를 추가해주면 된다.

```javascript
function QuestionPage() {
  const [popupOpened, setPopupOpened] = useState(false); // 팝업 상태
  async function handleQuestionSubmit() {
    // 연결 중인 전문가가 있으면 팝업 띄우기
    const 연결 전문가 = await 연결전문가_받아오기();
    if (연결 전문가 !== null){
      setPopupOpened(true);
    } else {
    const 약관동의 = await 약관동의_받아오기();
    if (!약관동의) {
      await 약관동의_팝업열기();
    }
    await 질문전송(questionValue);
    alert("질문이 등록되었어요.");
    }
  }

  // 팝업 버튼 클릭 핸들러
  async function handleMyExpertQuestionSubmit () {
    await 연결전문가_질문전송(questionValue, 연결전문가.id);
    alert(`${연결전문가.name}에게 질문이 등록되었어요`);
  }

  return (
    <>
    <form>
      <textarea placeholder="어떤내용이 궁금하신가요?" />
      <Button onClick={handleQuestionSubmit} />
    </form>
    //팝업 컴포넌트
    {popupOpened && (
      <연결 전문가 팝업 onSubmit={handleMyExpertQuestionSubmit} />
    )}
    </>
  );
}
```

타당하고 자연스러운 코드 추가처럼 보이지만 나쁜 코드가 되었다.

1. 하나의 목적인 코드가 흩뿌려져있다.
   => 코드가 뚝뚝 떨어져 있어서 나중에 기능을 추가할 때 스크롤을 위아래로 이동하며 미로 찾기를 해야한다.
2. 하나의 함수가 연결전문가, 약관 동의, ... , 여러 가지일을 수행한다.
   => 세부 구현을 모두 읽어야 함수의 역할을 이해할 수 있고, 코드 추가 및 삭제 시간이 더 오래 걸리게 된다.
3. 함수의 세부 구현 단계가 제각각이다.
   handleMyExpertQuestionSubmit 함수와 handleQuestionSubmit 함수 모두 이벤트 핸들링 관련 함수이다.

이렇게 기능을 추가하게 되면서 깔끔했던 코드가 작은 기능 하나를 추가했더니 어지러운 코드가 되었다.

#### 2-1. 큰 그림을 보며 코드 리팩토링

1. 함수 세부 구현 단계 통일
   함수들을 이름을 바꿔 위계를 맞춘다.

```javascript
function QuestionPage() {
  const 연결전문가 = useFetch(연결전문가_받아오기);
  async function handleNewExpertQuestionSubmit() {
    await 질문전송(questionValue);
    alert("질문이 등록되었어요.");
  }

  async function handleMyExpertQuestionSubmit() {
    await 연결전문가_질문전송(questionValue, 연결전문가.id);
    alert(`${연결전문가.name}에게 질문이 등록되었어요`);
  }
}
```

2. 하나의 목적인 코드는 뭉쳐두기
   기존에는 팝업을 여는 버튼과 팝업 코드가 동떨어져 있었는데 팝업 관련 코드를 하나로 모아 PopupTriggerButton 컴포넌트를 생성했다. 그리고 띄워줄 팝업을 prop으로 내려준다.

```javascript
return (
  <>
    <form>
      <textarea placeholder="어떤내용이 궁금하신가요?" />
      <Button onClick={handleQuestionSubmit} />

      {연결전문가.connected ? (
        <PopupTriggerButton
        popup={(
          <연결전문가팝업 onButtonSubmit={handleMyExpertQuestionSubmit} />
        )}>
        질문하기</PopupTriggerButton>
      )}
    </form>
  </>
);
```

3. 함수가 한 가지 일만 하도록 쪼개기

결론,
**클린 코드 != 짧은 코드 == 원하는 로직을 빠르게 찾을 수 있는 코드**

### 3. 로직을 빠르게 찾을 수 있는 코드

원하는 로직을 빠르게 찾으려면?

1. 하나의 목적을 가진 코드가 흩뿌려져 있으면 => 응집도
2. 함수가 여러 가지 일을 하고 있으면 => 단일 책임
3. 함수가 세부 구현 단계가 제각각이면 => 추상화

#### 3-1. 응집도 : 같은 목적의 코드는 뭉쳐 두자

하나의 기능인 팝업을 여는 기능이 코드가 여기 저기 흩뿌려져 있다.
=> 해당 코드는 파악도 한 번에 안되고 버그 발생 위헙도가 높은 코드이다.

- 리팩토링 버전 1
  해당 버전은 커스텀 훅을 사용해서 한 군데로 싹다 뭉쳐버렸다.

```javascript
function QuestionPage() {
  const [openPopup] = useMyExpertPopup(연결전문가.id);

  function handleClick() {
    openPopup();
  }

  return <button onClick={handleClick}>질문하기</button>;
}
```

커스텀 훅이 반환하는 openPopup만 호출만 해도 바로 팝업을 열 수 있다.

=> 그러면 문제가 발생한다!!
오히려 코드 파악하기가 어려워진다. 어떤 내용의 팝업을 띄우는지, 버튼 눌렀을 때 어떤 액션을 하는 지가 훅에 가려져버린다.

뭉치는 것도 무작정 같은 기능하는 코드를 한꺼번에 뭉치면 안된다!(**클린코드!=짧은코드**)

- 무엇을 뭉치지?
  뭉치면 쾌적 : 당장 몰라도 되는 디테일
  뭉치면 답답 : 코드 파악에 필수적인 핵심 정보

그렇다면 어떻게 응집해야 읽기 좋을까?
먼저 남겨야 할 핵심 데이터와 숨겨도 될 세부구현을 나눈다.
이 코드에서 **핵심 데이터**는 팝업 버튼 클릭 시 수행하는 액션과 팝업의 제목과 내용입니다.
또한, 팝업을 열고 닫을 때 사용하는 상태와 컴포넌트의 세부 마크업 그리고 팝업의 버튼 클릭 시 특정함수를 호출해야 한다는 바인딩이 **세부구현**에 해당한다.
![](https://velog.velcdn.com/images/hansoom3315/post/72a24bc5-2add-493f-9152-fee00bf5fc61/image.png)

- 리팩토링 버전 2
  핵심 데이터는 밖에서 전달하고, 나머지는 뭉친다!
  openPopup 커스텀 훅안에 모든 코드를 다 숨기는 것이 아니라, 세부 구현만 숨기고 핵심데이터인 팝업 제목이랑, 내용, 액션은 바깥에서 넘겨준다.
  => 세부 구현을 읽지 않고도 어떤 팝업인지 파악할 수 있다.

```javascript
function QuestionPage() {
  const [openPopup] = usePopup();
  async function handleClick() {
    const confirmed = await openPopup({
      title: "보험질문하기",
      content: <div>전문가가 설명드려오</div>,
    });
    if (confirmed) {
      await submitQuestion();
    }
  }

  async function submitQuestion(연결전문가) {
    await 질문전송(연결전문가.id);
    alert("질문을 전송했습니다.");
  }

  return <button onClick={handleClick}>질문하기</button>;
}
```

#### 선언적 프로그래밍

핵심 데이터만 전달하고 세부구현은 뭉쳐 숨겨 두는 개발 스타일
![](https://velog.velcdn.com/images/hansoom3315/post/5f6d53c6-4047-4682-aeb7-20b6bd087492/image.png)

제목, 내용, 질문 제출 선언을 하면 팝압이 이미 해둔 세부 구현을 바탕으로 해당 내용을 뿌리는 스타일이다.
즉, '무엇'을 하는 함수인지 빠른 이해가 가능하고, 세부 구현은 안쪽에 뭉쳐두어서 신경 쓸 필요가 없다. => 무엇 만 바꿔서 쉽게 재사용이 가능하다.

#### 명령형 프로그래밍

선언형으로 뭉쳐두지 않고, 하나하나 세부 구현을 작성하는 방식
=> 읽는 데 오래 걸리고 재사용이 어렵다.

선언적 코드가 무조건 좋은 건가? => 아니다! 두 방법 모두 상황에 맞게 유동적으로 사용하면 된다

#### 3-2. 단일 책임

하나의 일을 하는 뚜렷한 이름의 함수를 만들자

##### 예제1 - button

- before
  버튼 클릭 함수에 로그를 찍는 함수, API 콜이 섞여 있다.

```javascript
<button
  onClick={async () => {
    log("클릭");
    await openConfirm();
  }}
/>
```

- after
  LogClick 컴포넌트를 만들어 버튼을 감싸 클릭하면 자동으로 클릭 로그가 전송되도록 책임을 분리한다.

```javascript
<LogClick message="클릭">
  <button
    onClick={async () => {
      await openConfirm();
    }}
  />
</LogClick>
```

##### 예제2 - intersection observer

- before
  옵저버 세부 구현과 API콜이 함께 섞여 있는 코드이다.

```javascript
const target = useRef(null);

useEffect(() => {
  const observer = new IntersectionObserver(([{ isIntersecting }]) => {
    if (isIntersectiong) {
      fetchCats(nextPage);
    }
  });

  return () => {
    observer.unobserve(target.current);
  };
});

return <div ref={target}>더보기</div>;
```

- after
  Impression 옵저버 세부 구현은 감싼 컴포넌트에 숨겨두고, 사용하는 입장에서 Impression 시 API 콜만 신경 쓸 수 있다.

```javascript
<IntersectionArea onImpression={() => fetchCats(nextPage)}>
  더보기
</IntersectionArea>
```

#### 3-3. 추상화 : 핵심 개념 뽑아내기

##### 예제1 - 컴포넌트로 추상화

왼쪽은 팝업 컴포넌트 코드를 제로부터 디테일하게 구현
오른쪽은 팝업 코드를 제출 액션과 성공 액션이라는 중요한 개념만 남기고 나머지를 추상화
![](https://velog.velcdn.com/images/hansoom3315/post/2216a7db-04fd-4722-bbfc-a57520bd0e63/image.png)

##### 예제2 - 함수로 추상화

왼쪽은 전문가 정보를 받아와서 응답 값에 따라 다른 라벨을 보여주는 코드
오른쪽은 함수명 안에 모두 추상화 시킨 코드
![](https://velog.velcdn.com/images/hansoom3315/post/b8914d8a-7de1-428d-acd5-235c3c4403b6/image.png)

##### 얼마나 추상화

구체적인 코드를 조금 추상적이게 혹은 더욱 추상적이게 리팩토링 할 수 있다.

- level0
  버튼을 클릭하면 컨펌 창을 띄우고 컨펌 버튼을 클릭하면 특정 메시지를 띄우는 구체적인 코드이다.

```javascript
<Button onClick={showConfirm}>
  전송
</Button>
{isShowConfirm && (
  <Confirm onClick={() => showMessage("성공")}>
)}
```

- level1
  버튼을 눌렀을 때 컨펌창을 띄우는 기능을 컨펌 버튼이라는 컴포넌트로 추상화 => onConfirm 을 통해 원하는 액션을 넘겨줄 수 있다.

```javascript
<ConfirmButton onConfirm={() => showMessage("성공")}>전송</ConfirmButton>
```

- level2
  message 라는 prop만 넘겨서 컨펌창에 원하는 메시지를 보여줘 더 간단하게 추상화시킬 수 있다.

```javascript
<ConfirmButton message="성공">전송</ConfirmButton>
```

-level3
모든 기능을 컴포넌트 이름 아래에 추상화시킬 수도 있다.

```javascript
<ConfirmButton />
```

> 정답은 없다! 상황에 따라 필요한 만큼 추상화하면 된다.

추상화 수준이 섞여 있으면 코드 파악이 어렵다. 추상화 단계 높은 것끼리, 낮은 것끼리 모아서 추상화 단계를 비슷하게 정리해준다.
![](https://velog.velcdn.com/images/hansoom3315/post/714e6e69-db35-40a2-ac8b-172bc2fa8ce8/image.png)

### 4. 액션 아이템

1. 담대 하게 기존 코드 수정하기
   두려워 하지 말고 기존 코드를 씹고 뜯고 맛보고 즐기자
2. 큰 그림 보는 연습하기
   그 때는 맞고 지금은 틀리다.
   기능 추가 자체는 클린해도, 전체적으로 어지러울 수 있다.
3. 팀과 함께 공감대 형성하기
   코드에 정답은 없다! 명시적으로 이야기를 하는 시간이 필요하다.
4. 문서로 적어보기
   향후 어떤 점에서 위험할 수 있는 지 어떻게 개선할 수 있을 지를 글러 적어야 명확해 진다.
