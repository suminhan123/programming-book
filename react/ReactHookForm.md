## React Hook Form

React Hook Form 은 **비제어 컴포넌트로 렌더링을 최적화할 수 있는 라이브러리** 이다.

단순히 form을 처리하기 위해 state 로 모든 값을 검사하여 리랜더링 하는 것 보다 입력이 끝난 후 유효성 검사를 보여주어도 되고 더 빠른 검사를 할 수 있어 비제어 컴포넌트 방식인 React Hook Form을 많이 사용한다.

### ref 특징

ref 는 값을 업데이트하여도 리랜더링 되지 않는 특성으로, 입력이 모두 되고 난 후 ref 를 통해 값을 한번에 가져와서 활용한다.

state 로 값을 관리하지 않기 때문에 **값이 바뀔 때마다 리렌더링을 하지 않고** 값을 한 번에 가져올 수 있는 성능상에 이점이 있으나, 데이터를 완벽하게 가져올 수 없는 단점이 있다.

> 🔍 useRef() => heap 영역에 저장되는 자바스크립트 객체
>
> - 렌더링 할 때마다 동일한 객체를 제공한다. heap 에 저장을 하므로
>   어플리케이션이 종류되거나 가비지 컬렉팅이 되기 전까지 참조시에는 같은 메모리 값을 가진다.
>
> - 값이 변경이 되어도 리렌더링 되지 않는다. 같은 메모리 값을 항상 반환하므로 변경사항을 감지할 수 없어서 리렌더링을 하지 않는다.

#### React Hook Form 의 장점

**1. 간결한 API**
React Hook Form 은 사용하기 쉽고 직관적인 API를 제공하여 복잡한 폼 로직을 단순화 한다.
기본적으로 제공하는 Hook 함수들과 컴포넌트들을 사용하여 폼을 쉽게 생성하고 관리할 수 있다.

**2. 높은 성능**
React Hook Form 은 성능에 중점을 두어 최적화되어 있습니다.
입력 필드의 값 변화를 **추적하는 상태 대신 각 입력 필드의 참조를 사용**하여 불필요한 리렌더링을 방지하고, 가상 DOM의 업데이트를 최소화한다.

**3. 유효성 검사**
React Hook Form 은 내장된 유효성 검사를 지원하며, Yup, Joi 외부 유효성 검사 라이브러리와 통합할 수 있다.
입력 필드의 값에 대한 유효성 검사를 수행하고, 에러 메시지를 표시할 수 있다.

**4. 커스텀 훅**
React Hook Form 은 커스텀 훅을 사용하여 개발자가 필요한 로직을 쉽게 작성하고 재사용할 수 있도록 지원한다.
커스텀 훅을 사용하면 폼 상태, 에러 처리, 폼 제출 등의 로직을 캡슐화할 수 있다.

=> 개발하고 있는 서비스에는 많은 항목에 대한 입력을 받고 있기에 렌더링을 고려해 react hook form 을 활용해보고자 한다.

### React Hook Form 시작하기

```
yarn add react-hook-form
```

### React Hook Form 함수 살펴보기

```typescript
const {
    register,
    formState: { errors },
    handleSubmit,
    setError,
  } = useForm<IAuthForm>({mode: 'onBlur'});

const onSubmit = (data) => {
  console.log(data)
}

return (
  <form onSubmit={handleSubmit(onSubmit)>
  ...
```

#### register

input 요소를 React Hook Form 과 연결시켜 검증 규칙을 적용할 수 있게 하는 메소드

```typescript
const { register } = useForm();
const { name, ref, onChange, onBlur } = register("username");

<input
	name={name}
	ref={ref}
	onChange={onChange}
	onBlur={onBlur}
/>

// 객체 안에 value 를 일일이 하기 너무 많다
<input {...register("username")} />

```

register 가 리턴하는 객체 항목을 input 항목에 연결시키면 된다.

#### formState

form state 에 관한 정보를 담고 있는 객체

#### handleSubmit

form 을 submit 했을 때 실행할 함수
Validation을 통과했을 때 실행할 콜백함수가 반드시 필요하다. 실패했을 때의 콜백함수(submitErrorHandler)는 optional

#### setError

error 관련 설정에 사용되는 함수

#### mode

사용자가 submit 버튼을 누르기 전에 form 에 입력한 값이 유효한 값이 안라는 것을 미리 표시해주고 싶을 때 사용하는 것이 mode!

`mode` 는 useForm() 에 넘겨줄 수 있는 다양한 optional arguments 중 하나로 사용자가 form 을 submit 하기 전에 validation이 실행될 수 있게 해준다.

```typescript
// mode에 사용 가능한 값
mode: onChange | onBlur | onSubmit | onTouched | all = 'onSubmit'
```

### Devtool 설치하기

React Hook Form 은 Devtool 을 제공한다.
Form 관리를 일일이 콘솔 창에 출력하지 않고 현재 일어 나고 있는 상태를 쉽게 파악할 수 있다.

```
yarn add -D @hookform/devtools
```

### React Hook Form - FormProvider 생성하기

FormProvider 은 React Hook Form 에서 제공하는 컴포넌트로, React의 Context API를 기반으로 구현되었다.
FormProvider 은 상위 컴포넌트에서 하위 컴포넌트로 **폼 데이터와 관련된 상태와 로직을 전달**하기 위해 사용한다.

useFormContext 를 사용하면 컨텍스트를 prop으로 전달하는 것이 불편한 깊은 중첩 구조에서도 컨텍스트에 접근할 수 있다.

이 훅을 사용하면 useForm 에서 반환하는 모든 메서드와 속성을 가져올 수 있다. 즉, useForm 의 반환값을 그대로 사용할 수 있다.

![](https://velog.velcdn.com/images/hansoom3315/post/f2e1eb02-f07e-4e40-b1f3-ecc8b23bf076/image.png)

useFormContext 를 사용하기 위해서는 폼을 FormProvider 컴포넌트로 감싸줘야 한다.
FormProvider 컴포넌트에 useForm에서 반환한 메서드와 속성을 전달하면 된다.
그런 다음 useFormContext를 호출하면 해당 메서드와 속성을 가져올 수 있다.

### FormProvider 사용 방법

1. useForm() 훅을 사용하여 폼 메서드를 가져온다.

```typescript
import { useForm, FormProvider } from 'react-hook-form';
...
const methods = useForm();
...
```

2. FormProvider 의 props 로 1 에서 가져온 폼 메서드를 넘겨준다.

```typescript
<FormProvider {...methods}>
```

3. 폼 컨포넌트들을 FormProvider 내부에서 작성한다.
   폼 컴포넌트를 포함한 컴포넌트를 childComponent 로 사용하게 되면, 원래는 props를 넘겨줘야 하지만 FormProvider 로 감싼다면 따로 Props 를 넘기지 않아도 된다.

```typescript
<FormProvider {...methods}>
  //폼 컴포넌트가 들어있는 childComponent로
  <ChildComponent />
</FormProvider>
```

4. childComponent 내부에서 useFormContext 를 통해 useForm 의 반환값을 그대로 사용할 수 있다.
   여기서 props drilling 없이 사용이 가능하다.

```typescript
const ChildComponent = () => {
  const { register, handleSubmit } = useFormContext();
  const onSubmit = (data) => console.log(data);

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <input {...register("test")} />
      <input type="submit" />
    </form>
  );
};
```

#### 전체 코드

```typescript
import React from "react";
import { useForm, FormProvider, useFormContext } from "react-hook-form";

const ParentForm = () => {
  const methods = useForm();

  return (
    <FormProvider {...methods}>
      <ChildForm />
    </FormProvider>
  );
};

const ChildForm = () => {
  const { register, handleSubmit } = useFormContext();
  const onSubmit = (data) => console.log(data);

  return (
    <form onSubmit={methods.handleSubmit(onSubmit)}>
      <input type="text" name="firstName" {...methods.register("firstName")} />
      <input type="text" name="lastName" {...methods.register("lastName")} />
      <button type="submit">Submit</button>
    </form>
  );
};

const App = () => {
  return (
    <div>
      <h1>My Form</h1>
      <ParentForm />
    </div>
  );
};

export default App;
```

### FormProvider 의 렌더링

useFormContext 를 사용한다고 가정했을 때, FormProvider 의 자식 컴포넌트의 내부의 컴포넌트들의 예를 들어 input 이 수정하면 부모 컴포넌트들도 재렌더링이 될까? 아니다!

그 이유는 FormProvider 과 useFormContext 를 사용하는 핵심 개념인 로컬 폼 상태를 활용하여 최적화하였기 때문이다.

로컬 폼 상태는 해당 컴포넌트가 직접적으로 의존하는 폼 필드에 대해서만 관리하고 업데이트하므로, 다른 컴포넌트의 리렌더링에 영향을 받지 않는다.
=> 자식의 자식 컴포넌트들의 성능 이슈는 해당 필드의 변경에 따라 발생하며, 다른 부모나 형제 컴포넌트 들의 리렌더링 과는 독립적이다.

#### 주의해야 할 점

만약 부모 컴포넌트가 FormProvider 로 감싸져 있고, 그 하위에 여러개의 자식 컴포넌트가 있고, 그 자식 컴포넌트 중에서 useFormContext 를 사용하는 컴포넌트가 있다면, 해당 useFormContext 를 사용하는 컴포넌트의 로컬 폼 상태 변경이 발생하면 그 컴포넌트와 그의 하위 컴포넌트 들이 리렌더링 될 수 있다.
=> React 의 컴포넌트 트리에서 상위로의 리렌더링 전파로 인해 발생한다.

따라서, 깊은 계층 구조를 가지는 컴포넌트에서는 useFormContext 를 사용하는 컴포넌트의 성능에 영향을 주는 요소들을 최적화하는 것이 중요하다.

### React Hook Form 직접 적용하기

### FormProvider 적용

뽁에서는 많은 페이지에서 form data 를 다루기 때문에 FormProvider 을 활용해 context 로 데이터를 다른 컴포넌트에서도 prop전달없이 접근할 수 있도록 할 것이다.

```typescript
import { FormProvider, useForm } from "react-hook-form";

const WritingDiaryFormProvier = ({ children }: PropsWithChildren) => {
  const methods = useForm<IDiaryContextBody>();
  const onSubmit = (data: IDiaryContextBody) => {
    console.log(data);
  };
  const isMounted = useIsMounted();
  return (
    <FormProvider {...methods}>
      {isMounted && <DevTool control={methods.control} />}
      <form
        className="flex size-full flex-col"
        onSubmit={methods.handleSubmit(onSubmit)}
      >
        {children}
      </form>
    </FormProvider>
  );
};
export default WritingDiaryFormProvier;
```

사용자로 부터 입력을 받는 받는 페이지의 layout 으로 다음과 같이 컴포넌트로 감싸준다.

```typescript
"use client";

import { WritingDiaryFormProvider } from "@features/diary/contexts";
import usePreventLeave from "@hooks/usePreventLeave";
import { type PropsWithChildren } from "react";

const WritingLayout = ({ children }: PropsWithChildren) => {
  usePreventLeave();
  return <WritingDiaryFormProvider>{children}</WritingDiaryFormProvider>;
};
export default WritingLayout;
```

### usePreventLeave() 훅

여기서 페이지를 이탈하거나 혹은 새로고침을 할 경우 폼 데이터가 초기화하기에 사용자에게 이탈 전 아래와 같이 alert 창을 띄워주는 hook인 usePreventLeaver 를 추가해주었다.

훅은 weindow.onbeforeunload 에 beforeUnloadHandler 를 할당하여 페이지를 떠날 때마다 사용자에게 경고를 표시해준다.

```typescript
import { useRouter } from "next/navigation";
import { useEffect } from "react";

const usePreventLeave = () => {
  const router = useRouter();

  useEffect(() => {
    const beforeUnloadHandler = (e: BeforeUnloadEvent) => {
      e.preventDefault();
      e.returnValue = "";
    };
    window.onbeforeunload = beforeUnloadHandler;
    return () => {
      window.onbeforeunload = null;
    };
  }, [router]);
};

export default usePreventLeave;
```

### 다른 페이지에서 상태 업데이트

#### form element 가 아닌 경우

```typescript
const WritingEmojiForm = () => {
  const [selectEmoji, setSelectEmoji] = useState<TEmoji | null>(null);
  const { register, getValues, setValue, control } = useFormContext<IDiaryContextBody>();
  const { field } = useController({
    name: 'isChecked',
    control,
    defaultValue: true,
  });

  useEffect(() => {
    setSelectEmoji(getValues('emoji'));
  }, [getValues]);

  const handleSelectEmoji = (emoji: TEmoji) => {
    setSelectEmoji(emoji);
    setValue('emoji', emoji);
  };
  return (
    <>
      <h2 className="mb-3 text-base font-medium text-gray-65">감정</h2>
      <div className="flex justify-center gap-3" {...register('emoji')}>
        {DIARY_EMOJI_ARRAY.map((emoji) => (
          <Image
            className="cursor-pointer"
            loader={ImageLoader}
            width={40}
            height={40}
            key={emoji}
            src={emoji === selectEmoji ? DIARY_EMOJI[emoji].smallSelect : DIARY_EMOJI[emoji].smallNotSelect}
            onClick={() => handleSelectEmoji(emoji)}
            alt=""
          />
        ))}
      </div>
```

#### form element 인 경우

```typescript
const WritingDateForm = ({ defaultValue }: IWritingDateFormProp) => {
  const today = moment().format("YYYY-MM-DD");
  const { register, control } = useFormContext<IDiaryContextBody>();
  const { field } = useController({
    name: "date",
    control,
    defaultValue: defaultValue || today,
  });

  return (
    <>
      <h2 className="mb-3 mt-8 text-base font-medium text-gray-65">날짜</h2>
      <DatePicker
        {...register("date")}
        date={field.value}
        setDate={field.onChange}
      />
    </>
  );
};
```
