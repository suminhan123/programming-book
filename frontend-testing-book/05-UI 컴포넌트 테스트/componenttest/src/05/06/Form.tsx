import { useState } from "react";
import { ContactNumber } from "./ContactNumber";
import { DeliveryAddress } from "./DeliveryAddress";
import { PastDeliveryAddress } from "./PastDeliveryAddress";
import { RegisterDeliveryAddress } from "./RegisterDeliveryAddress";

/**
 * 테스트할 상황
 * 1. 이전 배송지가 없는 경우
 * 2. 이전 배송지가 있는데 새로운 배송지를 입력하지 않는 경우
 * 3. 이전 배송지가 있는데 새로운 배송지를 입력하는 경우
 */
export type AddressOption = React.ComponentProps<"option"> & { id: string };
export type Props = {
  deliveryAddresses?: AddressOption[];
  onSubmit?: (event: React.FormEvent<HTMLFormElement>) => void;
};
export const Form = (props: Props) => {
  const [registerNew, setRegisterNew] = useState<boolean | undefined>(
    undefined
  );
  return (
    <form onSubmit={props.onSubmit}>
      <h2>배송지 정보 입력</h2>
      <ContactNumber />
      {props.deliveryAddresses?.length ? (
        <>
          <RegisterDeliveryAddress onChange={setRegisterNew} />
          {registerNew ? (
            <DeliveryAddress title="새로운 배송지" />
          ) : (
            <PastDeliveryAddress
              disabled={registerNew === undefined}
              options={props.deliveryAddresses}
            />
          )}
        </>
      ) : (
        <DeliveryAddress />
      )}
      <hr />
      <div>
        <button>주문내용 확인</button>
      </div>
    </form>
  );
};
