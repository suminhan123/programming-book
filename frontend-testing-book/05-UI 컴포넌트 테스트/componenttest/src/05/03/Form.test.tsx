import { fireEvent, render, screen } from "@testing-library/react";
import { Form } from "./Form";

/**
 * render 함수를 사용해 매개변수 name 이 할다한 값이 제대로 표시되는 가 테스트
 */
test("이름을 표시한다", () => {
  render(<Form name="taro" />);
});

/**
 * 렌더링된 요소 중 특정 DOM 요소를 취득하려면 screen.getByText 를 사용
 * => 일치한 문자열을 가진 한 개의 텍스트 요소를 찾는 api
 */
test("이름을 표시한다", () => {
  render(<Form name="taro" />);
  console.log(screen.getByText("taro"));
});

/**
 * toBeInTheDocument() 는 해당 요소가 DOM에 존재하는 가를 검증하는 커스텀 매처
 * => props 에 넘겨준 이름이 표시되는 가를 테스트할 수 있다
 */
test("이름을 표시한다", () => {
  render(<Form name="taro" />);
  expect(screen.getByText("taro")).toBeInTheDocument();
});

/**
 * screen.getByRole 함수는 특정 DOM요소를 역할로 취득할 수 있다
 */
test("버튼을 표시한다", () => {
  render(<Form name="taro" />);
  expect(screen.getByRole("button")).toBeInTheDocument();
});

/**
 * heading 을 screen.getByRole 로 취득하면 h2 참조를 얻을 수 있다
 * h2에 원하는 문자가 포함됐는 지 테스트 가능
 */
test("heading을 표시한다", () => {
  render(<Form name="taro" />);
  expect(screen.getByRole("heading")).toHaveTextContent("계정 정보");
});

/**
 * 테스트 라이브러리는 암묵적 영학을 활용할 쿼리를 우선적으로 사용하도록 권장
 */

/**
 * 이벤트 핸들러 호출은 함수를 목함수로 검증
 * fireEvent 를 사용하면 임의의 DOM이벤트 발생이 가능
 */
test("버튼을 클릭하면 이벤트 핸들러가 실행된다", () => {
  const mockFn = jest.fn();
  render(<Form name="taro" onSubmit={mockFn} />);
  fireEvent.click(screen.getByRole("button"));
  expect(mockFn).toHaveBeenCalled();
});

/**
 * UI 컴포넌트가 예기치 않게 변경됐는 지 검증하고 싶다면 스냅숏 테스트
 * UI 컴포넌트를 HTML 문자열로 해당 시점의 렌더링 결과를 외부 파일에 저장
 */
test("Snapshot : 계정명인 'taro' 가 표시된다", () => {
  const { container } = render(<Form name="taro" />);
  expect(container).toMatchSnapshot();
});

/**
 * 회귀 테스트 발생시키기
 * 이미 커밋된 .snap 파일과 현시점의 스냅숏 파일을 비교하여 차이점이 발견되면 테스트가 실패
 */
test("Snapshot : 계정명인 'taro' 가 표시됐는 지 확인한다", () => {
  const { container } = render(<Form name="jiro" />);
  expect(container).toMatchSnapshot();
});

/**
 * 실패한 스냅숏을 갱신하기 위해
 * npx jest --updateSnapshot
 */
