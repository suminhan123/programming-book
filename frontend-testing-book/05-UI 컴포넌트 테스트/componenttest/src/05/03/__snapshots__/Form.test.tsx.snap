// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot : 계정명인 'taro' 가 표시됐는 지 확인한다 1`] = `
<div>
  <form>
    <h2>
      계정 정보
    </h2>
    <p>
      jiro
    </p>
    <div>
      <button>
        수정
      </button>
    </div>
  </form>
</div>
`;

